### PR TITLE
Gallery Grid Block

### PR DESCRIPTION
**Pull request by ScottGarrison**
## What does this do/fix?

Adds in the Gallery Grid Block to SQ1

- Gives the ability to pull images used into a slideshow
- Settings allow option to choose column layout 1 through 4 up.

### CODE REVIEW

File: wp-content/plugins/core/src/Blocks/Blocks_Definer.php
Comment (dpellenwood) [inline wp-content/plugins/core/src/Blocks/Blocks_Definer.php:33]: Would you kindly import (`use ...`) the namespace and then class the class name only here?

File: wp-content/plugins/core/src/Blocks/Types/Gallery_Grid/Gallery_Grid_Model.php
Comment (dpellenwood) [inline wp-content/plugins/core/src/Blocks/Types/Gallery_Grid/Gallery_Grid_Model.php:15]: Similar here, `(use ... )` and just the class name here.

File: wp-content/themes/core/assets/js/src/theme/core/components.js
Comment (dpellenwood) [inline wp-content/themes/core/assets/js/src/theme/core/components.js:14]: Let's keep the name name consistent here (`blockGalleryGrid` or similar), since there's a good chance we'll have other "gallery" modules to import.

File: wp-content/themes/core/components/blocks/gallery_grid/Gallery_Grid_Controller.php
Comment (dpellenwood) [inline wp-content/themes/core/components/blocks/gallery_grid/Gallery_Grid_Controller.php:207]: Can you adjust this to use a PHP array to define the options, then use `json_encode()` to encode them into the JSON string before sending it along? It makes reading or editing the options / config much easier later.  [Here's an example from Conroe](https://github.com/ConroeISD/external-website/blob/sprint/2020.11/wp-content/plugins/core/src/Templates/Panels/MediaGallery.php#L232-L244) (uses a different slider lib, but the concept is the same.)

File: wp-content/themes/core/components/blocks/gallery_grid/Gallery_Grid_Controller.php
Comment (dpellenwood) [inline wp-content/themes/core/components/blocks/gallery_grid/Gallery_Grid_Controller.php:227]: I think there's probably a way to simplify this method a bit, but I'll leave it to a BE to suggest any changes. ;)

File: wp-content/themes/core/components/blocks/gallery_grid/Gallery_Grid_Controller.php
Comment (dpellenwood) [inline wp-content/themes/core/components/blocks/gallery_grid/Gallery_Grid_Controller.php:269]: Hmm.... I'd prefer to see you use an actual image component here, along with a few image sizes (even just WP's `medium` & `large` in addition to the `CORE_FULL` size. That will better support mobile devices along with wrapping the slide content in a `figure` element.

Additionally, the Image component supports an `$html` property which can be used to add a `figcaption`, which is your HTML markup below.

File: wp-content/themes/core/components/blocks/gallery_grid/Gallery_Grid_Controller.php
Comment (dpellenwood) [inline wp-content/themes/core/components/blocks/gallery_grid/Gallery_Grid_Controller.php:272]: A few things here:
* Hard-coding the html markup here feels wrong. Would it be possible to use a container and text components here instead?
* For the count string, please use: `sprintf( __( '%d of %d', 'tribe' ), $index + 1, count( $this->get_gallery_img_ids() );` It's cleaner and easier to translate later.
* I think the `of` should be lower cased, even if it is styled differently in the designs.

File: wp-content/themes/core/components/blocks/gallery_grid/Gallery_Grid_Controller.php
Comment (dpellenwood) [inline wp-content/themes/core/components/blocks/gallery_grid/Gallery_Grid_Controller.php:315]: I worry that this, too, will be too large for mobile devices. I think we should use both a `src_size` size and `srcset_sizes` so we don't unintentionally load over-sizes images.

File: wp-content/themes/core/components/blocks/gallery_grid/css/dialog.pcss
Comment (dpellenwood) [inline wp-content/themes/core/components/blocks/gallery_grid/css/dialog.pcss:11]: Please use the existing spacer values for any padding/margins (here and throughout).

File: wp-content/themes/core/components/blocks/gallery_grid/css/dialog.pcss
Comment (dpellenwood) [inline wp-content/themes/core/components/blocks/gallery_grid/css/dialog.pcss:7]: Hmm.... this component's css is quite a collection of other component's styles. I think you need to more clearly define what is specific to a generic `dialog` and move the rest of the styles to other component's styles or in the case of the swiper styles, move those to that integration's styles.

File: wp-content/themes/core/components/blocks/gallery_grid/css/dialog.pcss
Comment (dpellenwood) [inline wp-content/themes/core/components/blocks/gallery_grid/css/dialog.pcss:32]: Does this class exist in the blocks/components system? If it does, it should be renamed.

File: wp-content/themes/core/components/blocks/gallery_grid/css/dialog.pcss
Comment (dpellenwood) [inline wp-content/themes/core/components/blocks/gallery_grid/css/dialog.pcss:130]: Please be sure to use existing color vars (here and throughout).

File: wp-content/themes/core/components/blocks/gallery_grid/css/dialog.pcss
Comment (dpellenwood) [inline wp-content/themes/core/components/blocks/gallery_grid/css/dialog.pcss:140]: This is unnecessarily high. Can we use something more reasonable?

File: wp-content/themes/core/components/blocks/gallery_grid/css/dialog.pcss
Comment (dpellenwood) [inline wp-content/themes/core/components/blocks/gallery_grid/css/dialog.pcss:139]: Curious why the overlay is 20px wider than 100%?

File: wp-content/themes/core/components/blocks/gallery_grid/css/dialog.pcss
Comment (dpellenwood) [inline wp-content/themes/core/components/blocks/gallery_grid/css/dialog.pcss:152]: Please create a var for this since the value is reused a few places. it's easier to adjust a single var than have to remember to update values in multiple locations that should match.

File: wp-content/themes/core/components/blocks/gallery_grid/css/dialog.pcss
Comment (dpellenwood) [inline wp-content/themes/core/components/blocks/gallery_grid/css/dialog.pcss:125]: Can you add some CSS doc blocks to these selectors (here and throughout)? It makes grepping the code easier later.

File: wp-content/themes/core/components/blocks/gallery_grid/css/dialog.pcss
Comment (dpellenwood) [inline wp-content/themes/core/components/blocks/gallery_grid/css/dialog.pcss:206]: Imho, I think this should be abstracted to a unique "close" action style that we can use throughout the system rather than specific to the dialog here.

File: wp-content/themes/core/components/blocks/gallery_grid/css/dialog.pcss
Comment (dpellenwood) [inline wp-content/themes/core/components/blocks/gallery_grid/css/dialog.pcss:267]: This can use a `grid-cols` var.

File: wp-content/themes/core/components/blocks/gallery_grid/css/gallery_grid.pcss
Comment (dpellenwood) [inline wp-content/themes/core/components/blocks/gallery_grid/css/gallery_grid.pcss:1]: Can you add CSS doc blocks to this file?

File: wp-content/themes/core/components/blocks/gallery_grid/css/gallery_grid.pcss
Comment (dpellenwood) [inline wp-content/themes/core/components/blocks/gallery_grid/css/gallery_grid.pcss:29]: Please go ahead and use CSS grid (and grid gutters) for these layouts. No need to use flexbox for grid layouts. :tada:

File: wp-content/themes/core/components/blocks/gallery_grid/gallery_grid.php
Comment (dpellenwood) [inline wp-content/themes/core/components/blocks/gallery_grid/gallery_grid.php:15]: Curious why you're wrapping the content block in another container? You can apply the `b-gallery-grid__header` class to the content block (also, would be good to make it a html `<header>` if it's not), and you can use the content blocks' CTA field as the slideshow button rather than appending it after the fact. (less markup for the same effect.)

File: wp-content/themes/core/components/blocks/gallery_grid/gallery_grid.php
Comment (dpellenwood) [inline wp-content/themes/core/components/blocks/gallery_grid/gallery_grid.php:48]: Should the dialog title be a heading tag? Additionally, I think this should all be moved to a dialog component's template, and controller and rendered via a `get_template_part()` rather than hard coded in this component's template.

File: wp-content/themes/core/components/blocks/gallery_grid/index.pcss
Comment (dpellenwood) [inline wp-content/themes/core/components/blocks/gallery_grid/index.pcss:10]: This should be imported as it's own component.

File: wp-content/themes/core/components/blocks/gallery_grid/js/gallery_grid.js
Comment (dpellenwood) [inline wp-content/themes/core/components/blocks/gallery_grid/js/gallery_grid.js:11]: This trigger is pretty generic. Either it should apply to all dialogs (and be init'd in the dialog component), or the trigger should be specific to the gallery grid block.

**Review by dpellenwood — CHANGES_REQUESTED** [review]
A ton of great work here, Scott. Thanks for tackling this block.

There are some aspects of the code that need to be pulled a part some though. Specifically, I think you'll want to introduce a true, independent `dialog` component and separate the Gallery Grid block's code from the Dialog & Slider code a bit. As it it is right now, the components are too interwoven to be reusable (or editable) outside of this block's context.  Happy to jump on a call and talk it out.

File: wp-content/themes/core/components/blocks/gallery_grid/Gallery_Grid_Controller.php
Comment (mmcachran) [reply wp-content/themes/core/components/blocks/gallery_grid/Gallery_Grid_Controller.php:227]: return ! empty( $this->gallery ) ? array_filter( wp_list_pluck( $this->gallery, 'id' ) ) : [];

**Comment by ScottGarrison** [comment]
@dpellenwood Large overhaul done on this based on your feedback. Please give another run through when you are available.

File: wp-content/themes/core/assets/css/src/theme/actions/_mixins.pcss
Comment (dpellenwood) [inline wp-content/themes/core/assets/css/src/theme/actions/_mixins.pcss:270]: Some white space issues here. A `gulp lint` in the theme directory before merging should ix this and any other small issues throughout.

File: wp-content/themes/core/components/blocks/gallery_grid/Gallery_Grid_Controller.php
Comment (dpellenwood) [inline wp-content/themes/core/components/blocks/gallery_grid/Gallery_Grid_Controller.php:371]: Are `medium` and `medium_large` actual WP Core sizes? I wasn't aware that they existed.

File: wp-content/themes/core/components/dialog/Dialog_Controller.php
Comment (dpellenwood) [inline wp-content/themes/core/components/dialog/Dialog_Controller.php:12]: Minor but there are some white space / alignment issues here that should be cleaned up before merging.

File: wp-content/themes/core/components/dialog/dialog.php
Comment (dpellenwood) [inline wp-content/themes/core/components/dialog/dialog.php:6]: This is fantastic, Scott. Thanks for such good documentation here.

File: wp-content/themes/core/components/slider/css/slider.pcss
Comment (dpellenwood) [inline wp-content/themes/core/components/slider/css/slider.pcss:141]: Will this value work okay on small screens? It doesn't create horizontal scroll bars by chance?

**Review by dpellenwood — APPROVED** [review]
Some minor clean up and a couple questions. Other wise g2g for me! :squid: :dash:

**Pull request merged**
